### DISCUSSION TITLE
Examples of application that has more than one exported component

### DISCUSSION THREAD

**Discussion by norricorp** [slint-ui/slint#4709] [Q&A] [answered]
So far my investigations have shown that multiple exports in a single slint! macro fail. Multiple slint! macros in main.rs also fail.

Multiple compile statements in build.rs fail with multiple .slint files (only the last line is recognised). 

There is the example for todo but that only has a single export and wraps the dialog inside a popupWindow which I don't want unless that is the only way of getting a dialog.

So is there code showing how to do this? Does #3217 work as it is still open.

**Comment by ogoffart** [comment]
Multiple slint! macro in main.rs should in principle work.

For example, this works fine:

```
use slint::slint;

slint!{
    export component A inherits Window {
        background: blue;
    }
}


slint!{
    export component B inherits Window {
        background: green;
    }
}

fn main() {
    A::new().unwrap().show().unwrap();
    B::new().unwrap().show().unwrap();
    slint::run_event_loop().unwrap();
}
```

It might be that in some case there could be duplicate symbols, then you can try to put the slint! macro in a module.

Or why was it not working?

**Comment by norricorp** [comment]
Hi Oliver,
So if I use multiple slint! macros then the second slint! has a build error of 
error[E0252]: the name `TextStyle` is defined multiple times

And because the component is not exported then that causes other problems. Of course they may be caused by the way I have structured it.

I do think it is fair to say that multiple exported components, whether as multiple slint! or multiple slint files that are refered to by compile are not easy. And I would have thought this was fundamental functionality.

If you wish I can set up a few gist to show the code approaches I gave taken. There is an excellent possibility I have completely misunderstood how to use the library.

Regards,
John

**Comment by norricorp** [comment]
Just tried running your code and it works. Changed one of the windows to a dialog and still worked. So I guess I am doing something wrong.

**Comment by norricorp** [comment]
Here is the weird thing.
I took your code above and gradually bought in my code. 
Adding import to each slint! - worked
Adding my dialog code to the B sloint! - worked
Added my window code to the A slint! - compile error - the name `TextStyle` is defined multiple times

Where has TextStyle come from? It is almost as though multiple simple slint! are allowed by multiple complex slint! are not.

Another change. From above a "complex" dialog and a complex window gave the Textstyle error. But a sinple dialog and complex window worked.

slint!{
	import { Button, VerticalBox, LineEdit, StandardButton } from "std-widgets.slint";

    export component A inherits Window {
	
		width: 200px;
		height: 250px;
		
		in-out property<int> counter: 42;
		callback request-increase-value();
	
		VerticalLayout {
			VerticalBox {
				Text {
					text: "Counter: " + root.counter;
				}
				Button {
					text: "Increase value";
					clicked => {
						root.request-increase-value();
					}
				}
			}
		}
	}

}


slint!{
	import { Button, VerticalBox, LineEdit, StandardButton } from "std-widgets.slint";

    export component B inherits Dialog {
		Rectangle {
			background: green;
			Text { text: "hello";}
		}
    }
}

fn main() {
    A::new().unwrap().show().unwrap();
    B::new().unwrap().show().unwrap();
    slint::run_event_loop().unwrap();
}

And this does not work (replacing B)

slint!{
	import { Button, VerticalBox, LineEdit, StandardButton } from "std-widgets.slint";

    export component B inherits Dialog {
		callback echo-dialog(string);

		dialog-text := LineEdit {
				font-size: 14px;
				width: 100px;
				height: 16px;
				placeholder-text: "Enter text here";
			}

		StandardButton { kind: ok; clicked => { root.echo-dialog(dialog-text.text); } }
		StandardButton { kind: cancel; }
    }
}

So adding either of the LineEdit or the buttons causes the error. If it is just the callback declaration and Text then works.

**Reply by tronical** [reply, marked as answer]
I think the `TextStyle` comes from within the style. Generally there's a chance that using multiple `slint!` macros within the same module may produces clashes. I think it's better to use separate modules to avoid this:

```rust
mod foo {
slint::slint!{ export component A { ... } }
}
mod bar {
slint::slint! { export component B { ... } }
}

fn main() {
    foo::A::new().unwrap().show().unwrap();
    bar::B::new().unwrap().show().unwrap();
    slint::run_event_loop().unwrap();
}
```

**Comment by norricorp** [comment]
Hi Simon,
thanks for this. That does work. My original code still fails to compile but at least it is now finding all of the exported components. 
Regards,
John

**Comment by norricorp** [comment]
As I mentioned, this works for including the slint in the main file but how would I have multiple .slint files?
If I list multiple slint_build::compile statements then only the last is recognised. Combining the statements into a ; separated list fails.
The slint file is expected to have top level components, not modules.
I ask this because the template for a new project uses the .slint file apporach (though of course the tutorial uses the single main.rs approach) and I guess multiple slint files is easier to re-use. But only if multiple components can be exported.

**Reply by Kestrel96** [reply]
Hey, @norricorp , did you find a way to solve this? I am stuck at trying to use multiple slint files with components and did not find anything helpful.

**Reply by ogoffart** [reply]
Since Sint 1.7 released last July, it is possible to export several component that inherits from Window from the main file, and use these.
Seee https://slint.dev/blog/slint-1.7-released#multi-window

**Reply by Kestrel96** [reply]
Thanks for quick reply. Just to make sure, this is currently the preferred way to export components that inherit from `Window`? 

I have tried adding my window component to `build.rs` but it did not build (and older answers confirmed that there is (was?) a problem with this) on Slint 1.8.
